### ISSUE TITLE
JSON parser needs to handle nested objects better

### ISSUE THREAD

**Issue by a2sheppy** [kengruven/strukt-bugs#24]
Currently, if JSON involves nested objects, the objects wind up treated as a single text entry within the parent JSON. For example:

```
{
  "home_library": {
    "owner": "Bob Smith",
    "contents": {
      "books": [
        "The Jungle Book": {
          "isbn": "273828134",
          "summary": "A boy, a bear, and a life in the jungle.",
          "author": "Rudyard Kipling"
        },
        "Dogland": {
          ...
        }
        ...
      ]
      ...
    },
    "school_library": {
      ...
    }
}
```

When you parse this JSON, you wind up with two columns: "home_library" and "school_library", each of which in turn contains the entire JSON of the objects within as a single cell. That's obviously not ideal.

What would be helpful would be the ability for the Parse JSON operation to:

* Specify the path of the object or sub-object to parse (such as "home_library.contents.books")
* Allow parsing into a hierarchical tree rather than a grid, so that each nested level of objects is an expandable folder revealing the fields within.

A real-life example of the type of file I am trying to work with can be found here: https://raw.githubusercontent.com/mdn/browser-compat-data/master/api/AudioBuffer.json. I was hoping to use Strukt to work with our browser compatibility database, which consists of a ton of JSON like this; it's why I bought it, in fact.

That same repository contains lots of examples to look at: https://github.com/mdn/browser-compat-data

**Comment by kengruven** [comment]
I agree! -- and I've heard the same feedback from other users.  Strukt was originally designed for tabular data, but I'm increasingly seeing people use it with nested data, like JSON.

> Specify the path of the object or sub-object to parse (such as "home_library.contents.books")

I can appreciate wanting to accomplish this task, but I'm a little hesitant to pursue this approach, for a couple reasons.

 1. It seems JSON-specific, and I want to build tools that can work on any kind of data.  JSON is an exchange format, and I want to parse this (and every other format) into objects, and then operate on those objects.
 2. Typing structured text (like "foo.bar.baz") into a Strukt flag would mean dealing with escaping, and syntax errors, and a whole mess of interface complexity I've been (mostly) avoiding so far.
 3. If you just want to extract deeply-nested JSON by typing a path expression, [jq](https://stedolan.github.io/jq/) is great at that.  I don't think I can compete with them here.

Strukt will probably have a tree query feature eventually, but I'm not yet clear on what form it might take.

> Allow parsing into a hierarchical tree rather than a grid, so that each nested level of objects is an expandable folder revealing the fields within.

Good news!  This has been on my list for a while, and I'm already working on it.

Returning semi-parsed data as JSON was always kind of a hack.  I didn't anticipate needing to process nested data much, and returning JSON snippets was the simplest thing to implement, but it's admittedly not the nicest user interface.

I don't have a timeline for this feature, but probably "more than days, less than weeks".

**Comment by a2sheppy** [comment]
Yeah, I hear you. It’s a big difference in concept and implementation, so it will take some time. I understand the need to make smart decisions to keep the app viable for both types of data. I’ll keep an eye on things to see if the app progresses in a direction that lets me do the stuff I need to do — I’m insanely desperate for a really excellent visual JSON editor for Mac. :)


Sheppy

On December 12, 2017 at 9:35:42 PM, Ken Harris (notifications@github.com) wrote:

Returning semi-parsed data as JSON was always kind of a hack. I didn't anticipate needing to process nested data much, and returning JSON snippets was the simplest thing to implement, but it's admittedly not the nicest user interface.

I don't have a timeline for this feature, but probably "more than days, less than weeks".
